### ISSUE TITLE
E721 possible false positive

### ISSUE THREAD

**Issue by JoseKilo** [PyCQA/pycodestyle#776]
I have this code that uses `inspect` to analyse the typing annotations of a function:

```python
import inspect
import typing


def foo(bar: typing.Optional[str]):
    pass


def main():
    signature = inspect.signature(foo)
    for name, parameter in signature.parameters.items():

        for arg in parameter.annotation.__args__:
            if arg == type(None):
                print("It's NoneType")
```

It produces: `E721 do not compare types, use 'isinstance()'` on line 14.

I don't think it's a valid use case for `isinstance`, since there is no instance, just the type.

**Comment by hoylemd** [comment]
I was under the impression that `None` is a singleton instance of the `NoneType` instance, which would make `isinstance()` semantically correct in your example.  However, I think it's also considered best practices to compare things to None with `if arg is None`, because of that singleton nature.

Obviously, I don't have all of your context though, so don't take my word as gospel.

**Comment by sigmavirus24** [comment]
How are people even annotating `NoneType` on their functions? It doesn't seem possible. To me, this sounds like a completely contrived example unless I'm missing something about introspecting the annotations.

**Comment by JoseKilo** [comment]
Thank you for the replies. The example is a reduced version from a more complex case, but the point is:

1) `typing.Optional[str]` becomes `typing.Union[str, NoneType]`
2) We analyse the Union components using `parameter.annotation.__args__`
3) in this case one of those components is `NoneType`

I'm trying to compare to `NoneType` (not `None`, since the annotations are types) and do something different in that case.

I could use `if issubclass(arg, type(None)):` but it feels a bit weird, since `NoneType` can't actually be extended.

**Comment by FichteFoll** [comment]
I wonder if that wouldn't just be a case for flake8's `noqa: E721` or whether `type(None)` would need to be special-cased since you can write `if arg == str:` just fine and there is no type literal for `NoneType`.

**Comment by sigmavirus24** [comment]
Yeah, regardless, I don't think this is an issue with pycodestyle. There are a few known problems with the check, specifically around ORMs. This is not the first time folks have wanted some kind of special case exclusion. One could also do `isinstance(None, arg)` but that's not entirely clear either. I agree with @FichteFoll that `# noqa` or wholesale ignoring the rule makes sense for your case.

**Comment by GoldsteinE** [comment]
There is another case.
```python
is_optional = type(None) in typing.get_args(Optional[Union[int, Literal['a']]])
```
There is literally no way to replace this typecheck with isinstance (`isinstance(None, (int, Literal['a']))` and `isinstance(None, Literal['a'])` will both raise an exception).
